pkg/rebuild/rebuild/rebuildremote.go

```diff
@@ -91,6 +93,110 @@ var rebuildContainerTpl = template.Must(
+//     we cannot refer to the user by name from one if it's created by the
+//     other. Creating a user at the top-level, recreating it in the proxy, and
+//     using the uid in the build satisfies our constraints.
+//   - Docker socket access: For the proxy to read and write to the docker
+//     socket, it needs to be part of the owning user (root) or group
+//     (host-defined "docker" group unknown within the container). Associating
+//     the proxy with a shadowed "docker" group does not seem to work. Changing
+//     ownership of the socket resolves this, albeit suboptimally.
+//   - Docker build: To ensure the proxy cert is trusted during the build, each
+//     execution (i.e. RUN instruction) must be patched to mount the proxy
+//     certificate and add truststore env vars. This is currently done
+//     lexically on the input Dockerfile but could be modified to operate on
+//     the lower-level build representation.
+//
+// TODO: Support IPv6.
+var proxyBuildTpl = template.Must(
+	template.New(
+		"proxy build",
+	).Funcs(template.FuncMap{
+		"join": func(sep string, s []string) string { return strings.Join(s, sep) },
+	}).Parse(
+		textwrap.Dedent(`
+				set -eux
+				docker run --volume=/workspace:/workspace gcr.io/cloud-builders/gsutil cp -P gs://{{.UtilPrebuildBucket}}/proxy /workspace
+				docker network create proxynet
+				useradd --system {{.User}}
+				uid=$(id -u {{.User}})
+				docker run --detach --name=proxy --network=proxynet --privileged --volume=/workspace/proxy:/workspace/proxy --volume=/var/run/docker.sock:/var/run/docker.sock --entrypoint /bin/sh gcr.io/cloud-builders/docker -euxc '
+					useradd --system --non-unique --uid '$uid' {{.User}}
+					chown {{.User}} /workspace/proxy
+					chown {{.User}} /var/run/docker.sock
+					su - {{.User}} -c "/workspace/proxy \
+						-verbose=true \
+						-http_addr=:{{.HTTPPort}} \
+						-tls_addr=:{{.TLSPort}} \
+						-ctrl_addr=:{{.CtrlPort}} \
+						-docker_addr=:{{.DockerPort}} \
+						-docker_socket=/var/run/docker.sock \
+						-docker_truststore_env_vars={{join "," .CertEnvVars}} \
+						-docker_network=container:build \
+						-docker_java_truststore=true"
+				'
+				proxyIP=$(docker inspect -f '{{printf "%s" "{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}"}}' proxy)
+				docker network connect cloudbuild proxy
+				docker run --detach --name=build --network=proxynet --entrypoint=/bin/sh gcr.io/cloud-builders/docker -c 'sleep infinity'
+				docker exec --privileged build /bin/sh -euxc '
+					iptables -t nat -A OUTPUT -p tcp --dport {{.HTTPPort}} -j ACCEPT
+					iptables -t nat -A OUTPUT -p tcp --dport {{.TLSPort}} -j ACCEPT
+					iptables -t nat -A OUTPUT -p tcp -m owner --uid-owner '$uid' -j ACCEPT
+					iptables -t nat -A OUTPUT -p tcp --dport 80 -j DNAT --to-destination '$proxyIP':{{.HTTPPort}}
+					iptables -t nat -A OUTPUT -p tcp --dport 443 -j DNAT --to-destination '$proxyIP':{{.TLSPort}}
```

Comment:
Is there any concern that folks will work around the proxy by using other ports (may not be an issue for your use case, but wanted to make sure what the threat vectors were here)?

Reply:
Absolutely. As you've noted before, I think the main challenge here will be SSH with DNS also presenting difficulty. Once those are dealt with, we plan on blocking all other egress via similar rules.

Reply:
Gotcha, blocking all other egress (once a few other pieces are in place) sgtm!